### PR TITLE
Vue: pretty-print binding syntax

### PR DESCRIPTION
**Pull request by azz**
For #2097, this is my **work-in-progress** attempt at pretty-printing the contents of binding attributes for Vue.

### CODE REVIEW

File: src/multiparser.js [hunk 1]
```diff
@@ -139,26 +149,60 @@ function fromHtmlParser2(path, options) {
...
+       * Vue binding sytax: JS expressions
+       * :class="{ 'some-key': value }"
+       * v-bind:id="'list-' + id"
+       * v-if="foo && !bar"
+       */
+      if (/(^v-)|:/.test(node.name) && !/^\w+$/.test(node.value)) {
+        return {
+          text: node.value,
+          options: {
+            parser: parseJavaScriptExpression,
```
Comment (azz) [inline src/multiparser.js]: We can use the custom parser API here!

File: src/printer-htmlparser2.js [hunk 2]
```diff
@@ -103,7 +111,15 @@ function printAttributes(attribs) {
         if (attribs[name] === "") {
           return name;
         }
-        return concat([name, '="', attribs[name], '"']);
+
+        // FastPath requires the next node to be a property of the current one
+        node._currentAttribute = {
```
Comment (azz) [inline src/printer-htmlparser2.js]: This is the least worst way of doing this that I could think of.

The _correct_ way would be to reshape the AST between parsing and printing, but I'm not sure if that would hurt performance?

File: src/multiparser.js [hunk 3]
```diff
@@ -139,26 +149,60 @@ function fromHtmlParser2(path, options) {
...
+            // Use singleQuote since HTML attributes use double-quotes.
+            // TODO(azz): We still need to do an entity escape on the attribute.
+            singleQuote: true
+          },
+          transformDoc: doc =>
+            util.hasNewlineInRange(node.value, 0, node.value.length)
+              ? doc
+              : docUtils.removeLines(doc)
```
Comment (vjeux) [inline src/multiparser.js]: why do this?

File: src/multiparser.js [hunk 4]
```diff
@@ -139,26 +149,60 @@ function fromHtmlParser2(path, options) {
...
 
+function parseJavaScriptExpression(text, parsers) {
+  // Force parsing as an expression
+  const ast = parsers.babylon(`(${text})`);
+  // Extract expression from the declaration
+  return {
+    type: "File",
+    program: ast.program.body[0].expression
+  };
+}
```
Comment (vjeux) [inline src/multiparser.js]: This is so nice!

File: src/printer-htmlparser2.js (hunk 2 above)
Comment (vjeux) [inline src/printer-htmlparser2.js]: We're already doing it for css, I would also do it for HTML. We haven't done much performance work so far and it's been reasonable.

I'd rather do the right thing for now and when we decide that perf is important, then measure and optimize the things that are actually taking a lot of time and make trade-offs at that point.

Comment (vjeux) [inline src/printer-htmlparser2.js]: Adding it to the ast in this way is reasonable to me by the way, if you move it to the parsing phase, I'll be happy with it.

**Comment by vjeux** [comment]
Sounds good to me, I only have 2 comments.

File: src/multiparser.js (hunk 3 above)
Comment (azz) [inline src/multiparser.js]: Stopgap to remove the trailing hardline.

**Comment by vjeux** [comment]
Can't wait to land this :)

File: src/multiparser.js [hunk 5]
```diff
@@ -139,26 +149,60 @@ function fromHtmlParser2(path, options) {
...
     }
+
+    case "attribute-value": {
+      /*
+       * Vue binding sytax: JS expressions
+       * :class="{ 'some-key': value }"
+       * v-bind:id="'list-' + id"
+       * v-if="foo && !bar"
+       */
+      if (/(^v-)|:/.test(node.name) && !/^\w+$/.test(node.value)) {
```
Comment (lukepolo) [inline src/multiparser.js]: you may want to also include the `@` , for things like `@click="someFunction()"`

**Pull request merged**